Add NoUselessElseFixer

SpacePossum opened the pull request:
todo:
integration tests

replaces:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1693

keradus commented:
Also, Travis is red

SpacePossum commented:
that is because f the TODO, the integration tests are not done, just stubbed :)

keradus commented:
OK, cool ;)

gharlan commented on Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    /**
+     * Return the first and lost token index of the previous block.
```
"last"

gharlan commented:
Is there a test case similar to this?

```php
if ($x) {
    if ($y) {
        return 1;
    } else {
        return 2;
    }
} else {
    return 3;
}
```

SpacePossum commented:
TODO:
HHVM 3.6.1 https://3v4l.org/U2hoG tokenizes diff. than 3.8.1 / PHP Zend

keradus commented:
Ref https://github.com/keradus/PHP-CS-Fixer/commit/c6761cd74c6a8ac611936b9abdab6785217674a2

SpacePossum commented:
Should stay @ contrib level. I think this one is ready, please play around with it :)

keradus commented on Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php:
```diff
@@ -0,0 +1,217 @@
+        $tokens = Tokens::fromCode($content);
+
+        // start from 4; minimum number of tokens that form a candidate for fixing; <?php if ( [] ) [5]
+        foreach ($tokens as $index => $token) {
```
either fix the loop to respect the comment or remove the misleading comment

keradus commented:
Could anyone handle the review?

gharlan commented:
https://github.com/SpacePossum/PHP-CS-Fixer/pull/11

keradus commented:
@SpacePossum could you squash commits of yours into one and then leave @gharlan at the end?

SpacePossum commented:
>@SpacePossum could you squash commits of yours into one and then leave @gharlan at the end?

Sure thing :)
I'll work on this a bit more, little refactor, and than squash after the review.

keradus commented:
Oh, I thought it's ready to merge.
When done please remove WIP label

SpacePossum commented:
Happy to take review comments and thoughts on this one to wrap up it up :)
ping @gharlan and @keradus and everyone else.

keradus commented on Symfony/CS/Tests/AbstractIntegrationTest.php:
```diff
@@ -180,7 +180,8 @@ protected function doTestIntegration(IntegrationCase $case)
 
-        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $case->getFixers(), false, true, new FileCacheManager(false, null, $case->getFixers()));
+        $fixers = $case->getFixers();
+        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $case->getFixers(), false, true, new FileCacheManager(false, null, $fixers));
```
$case->getFixers() => $fixers

SpacePossum commented:
updated and squashed

keradus commented on Symfony/CS/Fixer/Contrib/NoUselessElseFixer.php:
```diff
@@ -0,0 +1,225 @@
+/**
+ * @author SpacePossum
+ */
+final class NoUselessElseFixer extends AbstractAlignFixer
```
```diff
-AbstractAlignFixer
+AbstractFixer
```

SpacePossum replied:
:+1:  done

GrahamCampbell commented:
Missing tests for else with colons?

SpacePossum commented:
wish they killed the alternative in PHP7... oh well.. lets see...

GrahamCampbell commented:
#PHP8 :P

SpacePossum commented:
>Missing tests for else with colons?

Test and a minor speed gain added. I didn't put effort in detecting / cleanup useless else statements done in the alternative syntax because it is no fun and an edge case anyway.

gharlan commented on Symfony/CS/Tests/AbstractIntegrationTest.php:
```diff
@@ -180,7 +180,8 @@ protected function doTestIntegration(IntegrationCase $case)
 
-        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $case->getFixers(), false, true, new FileCacheManager(false, null, $case->getFixers()));
+        $fixers = $case->getFixers();
+        $changed = $fixer->fixFile(new \SplFileInfo($tmpFile), $fixers, false, true, new FileCacheManager(false, null, $fixers));
```
Unrelated to this pr, right?

SpacePossum replied:
indeed, however I think it is of value, should I take it out?

gharlan replied:
In my opinion it should be another pr, together with the change in `IntegrationCase`.

SpacePossum commented:
Updated, rebased and clean upped.
The commit history looks a bit odd now with @gharlan being the author of the whole test file, hope he doesn't mind.

keradus commented:
Please, could you recommit to retrigger Travis? Also, put no_useless_else to .php_cs file

GrahamCampbell commented:
@keradus Or just close and re-open the PR.

keradus commented:
Didn't know that trick, thanks ;)
(But it's sad we need to have tricks for Travis :/)

gharlan commented:
@keradus you should be able to restart builds because you have write permissions to the repo.

![screenshot 2016-05-23 21 45 41](https://cloud.githubusercontent.com/assets/330436/15482409/c36e7b92-212f-11e6-98df-4f7198a85d81.png)

keradus commented:
@gharlan usually it is like that and I do it often, sadly this time it wasn't about crashed build, but the build was not triggered at all, so I have no build to restart  - I needed to retrigger it ;)

keradus commented:
Thank you @SpacePossum.

keradus commented:
Thank you @gharlan

SpacePossum commented:
: D
thanks @gharlan for additional test and reviews, @keradus and @GrahamCampbell thanks for reviewing  :+1:

localheinz commented:
@SpacePossum 

👍 Great fixer, but you wouldn't believe the code that was touched by it - I'm looking at it right now, and my eyes are bleeding! Nice job!
